src/memray/commands/attach.py

```diff
@@ -0,0 +1,310 @@
+import argparse
+import contextlib
+import os
+import pathlib
+import shutil
+import signal
+import socket
+import subprocess
+import threading
+
+import memray
+
+from .live import LiveCommand
+from .run import _get_free_port
+
+GDB_SCRIPT = pathlib.Path(__file__).parent / "_attach.gdb"
+RTLD_DEFAULT = memray._memray.RTLD_DEFAULT
+RTLD_NOW = memray._memray.RTLD_NOW
+PAYLOAD = """
+import atexit
+
+import memray
+
+
+def deactivate_last_tracker():
+    if hasattr(memray, "_last_tracker"):
+        tracker = memray._last_tracker
+        memray._last_tracker = None
+        if tracker:
+            tracker.__exit__(None, None, None)
+
+
+if not hasattr(memray, "_last_tracker"):
+    # This only needs to be registered the first time we attach.
+    atexit.register(deactivate_last_tracker)
+
+deactivate_last_tracker()
+
+tracker = {tracker_call}
+try:
+    tracker.__enter__()
+except:
+    # Prevent the exception from keeping the tracker alive.
+    # This way resources are cleaned up ASAP.
+    del tracker
+    raise
+
+memray._last_tracker = tracker
+"""
+
+# TODO:
+# - Allow the user to select between file vs socket (who picks the filename?).
+
+
+def inject(debugger: str, pid: int, port: int) -> bool:
+    """Executes a file in a running Python process."""
+    injecter = pathlib.Path(memray.__file__).parent / "_inject.abi3.so"
+    assert injecter.exists()
+
+    gdb_cmd = [
+        "gdb",
+        "-batch",
+        "-p",
+        str(pid),
+        "-nx",
+        "-nh",
+        "-nw",
+        "-iex=set auto-solib-add off",
+        f"-ex=set $rtld_now={RTLD_NOW}",
+        f'-ex=set $libpath="{injecter}"',
+        f"-ex=set $port={port}",
+        f"-x={GDB_SCRIPT}",
+    ]
+
+    # When adding new breakpoints, also update _attach.gdb
+    lldb_breakpoint_args = [
+        "breakpoint",
+        "set",
+        "-b",
+        "malloc",
+        "-b",
+        "calloc",
+        "-b",
+        "realloc",
+        "-b",
+        "free",
+        "-b",
+        "PyMem_Malloc",
+        "-b",
+        "PyMem_Calloc",
+        "-b",
+        "PyMem_Realloc",
+        "-b",
+        "PyMem_Free",
+        "-C",
+        "'breakpoint disable'",
+        "-C",
+        f"""'expr void *$dlopen = (void*)dlsym((void*){RTLD_DEFAULT}, "dlopen")'""",
+        "-C",
+        f"'expr ((void*(*)(const char*, int))$dlopen)($libpath, {RTLD_NOW})'",
+        "-C",
+        "'p (char*)dlerror()'",
+        "-C",
+        "'expr (const char*)memray::spawn_client($port)'",
+    ]
+
+    lldb_cmd = [
```

Comment:
`lldb` fails to execute this on MacOS because it demands that we cast these symbols to the correct versions. This patch makes it work:

```diff
diff --git a/src/memray/commands/attach.py b/src/memray/commands/attach.py
index 1e6f653..e14803a 100644
--- a/src/memray/commands/attach.py
+++ b/src/memray/commands/attach.py
@@ -112,13 +112,13 @@ def inject(debugger: str, pid: int, port: int) -> bool:
         "--no-lldbinit",
         "--source-quietly",
         "-o",
-        "p PyMem_Malloc",
+        "p ((void*(*)(size_t))PyMem_Malloc)",
         "-o",
-        "p PyMem_Calloc",
+        "p ((void*(*)(size_t, size_t))PyMem_Calloc)",
         "-o",
-        "p PyMem_Realloc",
+        "p ((void*(*)(void *, size_t))PyMem_Realloc)",
         "-o",
-        "p PyMem_Free",
+        "p ((void(*)(void*))PyMem_Free)",
         "-o",
         f'expr char $libpath[]="{injecter}"',
         "-o",
```